android/app/src/main/kotlin/io/lantern/model/SessionModel.kt

```diff
@@ -396,28 +401,53 @@ class SessionModel internal constructor(
                 val start = System.currentTimeMillis()
                 val appsList = appsDataProvider.listOfApps()
                 // First add just the app names to get a list quickly
-                val apps = buildJsonArray {
-                    appsList.forEach { app ->
-                        add(
-                            buildJsonObject {
-                                val byte = app.icon!!
-                                put("packageName", app.packageName)
-                                put("name", app.name)
-                                putJsonArray("icon") {
-                                    byte.toUByteArray().forEach { add(it.toInt()) }
+                val file = File(activity.cacheDir, "appsData.json")
+
+
+                // Write JSON data directly to the file to reduce memory usage
+                file.outputStream().use { outputStream ->
+                    Json.encodeToStream(buildJsonArray {
+                        appsList.forEach { app ->
+                            add(
+                                buildJsonObject {
+                                    val byte = app.icon ?: ByteArray(0)  // Safeguard for null icons
```

Comment:
I think adding this safeguard is useful. What will the user see when an icon is missing? Blank space? Do we want to display a generic app icon as a placeholder?

Reply:
I don't think that is possible, Most of them have an app Icon, Anyway, I am going to add some placeholder on the Flutter end.